File: pyJianYingDraftServer/app/routers/coze.py
"""
Coze API 路由
提供工作流、任务管理和插件数据接收功能
"""

from fastapi import APIRouter, HTTPException, Query
from typing import Optional

from app.models.coze_models import (
    CozeSendDataRequest,
    CozeSendDataResponse,
    Task,
    CreateTaskRequest,
    UpdateTaskRequest,
    ExecuteTaskRequest,
    ExecuteTaskResponse,
    TaskListResponse,
    TaskStatistics,
    TaskStatus,
    ExecutionStatus,
)
from app.services.coze_service import get_coze_service
from app.services.coze_workflow_service import get_workflow_service
from app.services.coze_client import get_coze_client

router = APIRouter()


# ==================== 插件数据接收（保留原有功能） ====================

@router.post("/send-data", response_model=CozeSendDataResponse, summary="Coze插件数据接收")
async def send_coze_data(request: CozeSendDataRequest):
    """
    接收来自Coze插件的数据

    - **api_base**: Coze插件的API基础URL
    - **client_id**: 客户端唯一标识
    - **data**: 要发送的数据，包含type和data字段

    数据接收后会通过WebSocket转发给对应clientId的订阅客户端
    """
    try:
        coze_service = get_coze_service()
        response = await coze_service.receive_data(request)
        return response

    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"处理Coze数据时发生错误: {str(e)}"
        )


@router.get("/stats", summary="获取Coze服务统计信息")
async def get_coze_stats():
    """
    获取Coze服务的统计信息

    返回包括：
    - 总订阅数
    - 总接收数据数
    - 总转发数据数
    - 当前活跃连接数
    """
    try:
        coze_service = get_coze_service()
        stats = coze_service.get_stats()

        return {
            "success": True,
            "data": stats,
            "message": "获取统计信息成功"
        }

    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"获取统计信息时发生错误: {str(e)}"
        )


@router.get("/health", summary="Coze服务健康检查")
async def health_check():
    """检查Coze服务是否正常运行"""
    try:
        coze_service = get_coze_service()
        stats = coze_service.get_stats()

        return {
            "status": "healthy",
            "service": "coze",
            "stats": stats
        }

    except Exception as e:
        raise HTTPException(
            status_code=503,
            detail=f"Coze服务不可用: {str(e)}"
        )


# ==================== 工作空间管理 ====================

@router.get("/workspaces", summary="获取工作空间列表")
async def get_workspaces(account_id: str = Query(default="default", description="账号ID")):
    """
    获取工作空间列表

    - **account_id**: 账号ID（默认为 default）
    """
    try:
        client = get_coze_client(account_id)
        if not client:
            raise HTTPException(
                status_code=400,
                detail=f"无法获取 Coze 客户端（账号: {account_id}），请检查配置"
            )

        workspaces = await client.list_workspaces()

        return {
            "success": True,
            "workspaces": workspaces,
            "count": len(workspaces)
        }

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"获取工作空间列表失败: {str(e)}"
        )


# ==================== 工作流管理 ====================

@router.get("/workflows/{workflow_id}", summary="获取工作流详情")
async def get_workflow(
    workflow_id: str,
    account_id: str = Query(default="default", description="账号ID")
):
    """
    获取工作流详情

    - **workflow_id**: 工作流ID
    - **account_id**: 账号ID
    """
    try:
        client = get_coze_client(account_id)
        if not client:
            raise HTTPException(
                status_code=400,
                detail=f"无法获取 Coze 客户端（账号: {account_id}），请检查配置"
            )

        workflow = await client.retrieve_workflow(workflow_id)

        return {
            "success": True,
            "workflow": workflow
        }

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"获取工作流详情失败: {str(e)}"
        )


@router.get("/workflows/{workflow_id}/history", summary="获取工作流执行历史")
async def get_workflow_history(
    workflow_id: str,
    account_id: str = Query(default="default", description="账号ID"),
    page_size: int = Query(default=20, ge=1, le=100, description="每页数量"),
    page_index: int = Query(default=1, ge=1, description="页码")
):
    """
    获取工作流执行历史

    - **workflow_id**: 工作流ID
    - **account_id**: 账号ID
    - **page_size**: 每页数量（1-100）
    - **page_index**: 页码（从1开始）
    """
    try:
        client = get_coze_client(account_id)
        if not client:
            raise HTTPException(
                status_code=400,
                detail=f"无法获取 Coze 客户端（账号: {account_id}），请检查配置"
            )

        history = await client.get_execution_history(
            workflow_id=workflow_id,
            page_size=page_size,
            page_index=page_index
        )

        return {
            "success": True,
            **history
        }

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"获取执行历史失败: {str(e)}"
        )


# ==================== 任务管理 ====================

@router.get("/tasks", response_model=TaskListResponse, summary="获取任务列表")
async def get_tasks(
    workflow_id: Optional[str] = Query(default=None, description="工作流ID筛选"),
    status: Optional[TaskStatus] = Query(default=None, description="任务状态筛选"),
    execution_status: Optional[ExecutionStatus] = Query(default=None, description="执行状态筛选"),
    limit: int = Query(default=50, ge=1, le=200, description="限制返回数量"),
    offset: int = Query(default=0, ge=0, description="偏移量")
):
    """
    获取任务列表，支持筛选和分页

    - **workflow_id**: 按工作流ID筛选
    - **status**: 按任务状态筛选
    - **execution_status**: 按执行状态筛选
    - **limit**: 限制返回数量（1-200）
    - **offset**: 偏移量
    """
    try:
        service = get_workflow_service()
        result = await service.list_tasks(
            workflow_id=workflow_id,
            status=status,
            execution_status=execution_status,
            limit=limit,
            offset=offset
        )

        return TaskListResponse(**result)

    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"获取任务列表时发生错误: {str(e)}"
        )


@router.get("/tasks/statistics", response_model=TaskStatistics, summary="获取任务统计")
async def get_task_statistics(
    workflow_id: Optional[str] = Query(default=None, description="工作流ID筛选")
):
    """
    获取任务统计信息

    - **workflow_id**: 按工作流ID筛选统计
    """
    try:
        service = get_workflow_service()
        stats = await service.get_task_statistics(workflow_id=workflow_id)

        return TaskStatistics(**stats)

    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"获取任务统计时发生错误: {str(e)}"
        )


@router.get("/tasks/{task_id}", response_model=Task, summary="获取单个任务")
async def get_task(task_id: str):
    """获取指定任务的详细信息"""
    try:
        service = get_workflow_service()
        task = await service.get_task(task_id)

        if not task:
            raise HTTPException(
                status_code=404,
                detail=f"任务 {task_id} 不存在"
            )

        return task

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"获取任务时发生错误: {str(e)}"
        )


@router.post("/tasks", response_model=Task, summary="创建任务")
async def create_task(
    request: CreateTaskRequest,
    account_id: str = Query(default="default", description="账号ID")
):
    """
    创建新任务（不执行）

    - **name**: 任务名称（可选，默认使用工作流名称）
    - **description**: 任务描述
    - **workflow_id**: 工作流ID
    - **workflow_name**: 工作流名称（可选）
    - **input_parameters**: 输入参数
    - **tags**: 标签
    - **priority**: 优先级
    - **metadata**: 扩展元数据
    """
    try:
        service = get_workflow_service()
        task = await service.create_task(request, account_id=account_id)

        return task

    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"创建任务时发生错误: {str(e)}"
        )


@router.put("/tasks/{task_id}", response_model=Task, summary="更新任务")
async def update_task(task_id: str, request: UpdateTaskRequest):
    """
    更新任务信息

    支持部分更新，只更新提供的字段
    """
    try:
        service = get_workflow_service()
        task = await service.update_task(task_id, request)

        return task

    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"更新任务时发生错误: {str(e)}"
        )


@router.delete("/tasks/{task_id}", summary="删除任务")
async def delete_task(task_id: str):
    """删除指定任务"""
    try:
        service = get_workflow_service()
        success = await service.delete_task(task_id)

        if not success:
            raise HTTPException(
                status_code=404,
                detail=f"任务 {task_id} 不存在"
            )

        return {
            "success": True,
            "message": f"任务 {task_id} 已删除"
        }

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"删除任务时发生错误: {str(e)}"
        )


@router.post("/tasks/execute", response_model=ExecuteTaskResponse, summary="执行任务")
async def execute_task(
    request: ExecuteTaskRequest,
    account_id: str = Query(default="default", description="账号ID")
):
    """
    执行工作流任务（真实调用 Coze API）

    - **task_id**: 任务ID（可选，不携带则生成新任务）
    - **workflow_id**: 工作流ID
    - **input_parameters**: 输入参数（可选，使用任务存储的参数）
    - **save_as_task**: 是否保存为任务（默认true）
    - **task_name**: 任务名称（保存为任务时使用）
    - **task_description**: 任务描述

    **注意**: 此接口会真实调用 Coze API 执行工作流，并自动管理任务状态
    """
    try:
        service = get_workflow_service()
        response = await service.execute_task(request, account_id=account_id)

        return response

    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"执行任务时发生错误: {str(e)}"
        )

File: pyjianyingdraft-web/src/types/coze.ts
// Coze Zone 相关类型定义

export interface CozeAccount {
  id: string;
  name: string;
  apiKey: string;
  baseUrl?: string;
  isActive: boolean;
  description?: string;
  createdAt: string;
  updatedAt: string;
}

export interface CozeWorkspace {
  id: string;
  name: string;
  description?: string;
  icon?: string;
  created_time: string;
  updated_time: string;
}


// Coze API 返回的工作流详细信息结构
export interface CozeWorkflowDetail {
  workflow_name: string;
  app_id: string;
  creator: {
    id: string;
    name: string;
  };
  updated_at: number; // Unix 时间戳
  workflow_id: string;
  description?: string;
  icon_url?: string;
  created_at: number; // Unix 时间戳
}

export interface CozeWorkflowInput {
  parameters: Record<string, {
    type: string;
    required?: boolean;
    description?: string;
    default_value?: any;
    items?: any; // 当类型为 array 时定义子类型
    properties?: Record<string, any>; // 当类型为 object 时定义属性
  }>;
}

export interface CozeWorkflowOutput {
  parameters?: Record<string, {
    type: string;
  }>;
  terminate_plan?: 'return_variables' | 'use_answer_content';
  content?: string; // 当 terminate_plan 为 use_answer_content 时
}

export interface CozeWorkflowInfo {
  workflow_detail: CozeWorkflowDetail;
  input?: CozeWorkflowInput;
  output?: CozeWorkflowOutput;
}

export interface CozeWorkflowApiResponse {
  data: CozeWorkflowInfo;
  code: number;
  msg: string;
  detail: {
    logid: string;
  };
}

// 兼容性接口 - 内部使用的格式
export interface CozeWorkflow {
  id: string;
  name: string;
  description?: string;
  created_time: string;
  updated_time: string;
  version: number;
  input_schema?: any;
  output_schema?: any;
  status: 'active' | 'inactive' | 'draft';
  workflow_definition?: any;
  publish_status?: string;

  // 详细信息字段（从API获取）
  icon_url?: string;
  app_id?: string;
  creator_id?: string;
  creator_name?: string;
  category?: string;
  tags?: string[];
  template_id?: string;
  is_template?: boolean;
  use_count?: number;
  like_count?: number;
  publish_time?: string;
  last_used_time?: string;
  node_count?: number;
  connection_count?: number;
  workflow_config?: {
    timeout?: number;
    max_retry_times?: number;
    variables?: Record<string, any>;
  };
}

export interface WorkflowExecution {
  id: string;
  workflow_id: string;
  workflow_name?: string;
  status: 'running' | 'success' | 'failed' | 'cancelled';
  input_data: any;
  output_data?: any;
  error_message?: string;
  created_time: string;
  completed_time?: string;
  duration?: number;
}

export interface CozeZoneTabData {
  id: string;
  label: string;
  type: 'coze_zone';
  loading: boolean;
  error: string | null;

  // Coze Zone 特有字段
  accountId: string;
  workspaceId: string;

  // 数据
  accounts: CozeAccount[];
  workspaces: CozeWorkspace[];
  workflows: CozeWorkflow[];
  executions: WorkflowExecution[];
  selectedWorkflow?: CozeWorkflow;
  executionHistory: WorkflowExecution[];

  // 状态
  refreshing: boolean;
  executing: boolean;
}

// Coze API 请求和响应类型
export interface CreateWorkspaceRequest {
  name: string;
  description?: string;
  icon?: string;
}


export interface ExecuteWorkflowRequest {
  workflow_id: string;
  parameters?: Record<string, any>;
  stream?: boolean;
  conversation_id?: string;
  bot_id?: string;
  user_id?: string;
}

export interface ExecuteWorkflowResponse {
  conversation_id: string;
  status: string;
  data: {
    id: string;
    workflow_id: string;
    status: string;
    input_data: any;
    output_data?: any;
    created_time: string;
  };
}

export interface CozeApiResponse<T = any> {
  code: number;
  message: string;
  data: T;
  error?: {
    code: number;
    message: string;
    details?: any;
  };
}

// Coze 错误码
export enum CozeErrorCode {
  // 通用错误
  SUCCESS = 0,
  INVALID_REQUEST = 4000,
  UNAUTHORIZED = 4010,
  FORBIDDEN = 4030,
  NOT_FOUND = 4040,
  RATE_LIMITED = 4290,
  INTERNAL_ERROR = 5000,

  // 工作流错误
  WORKFLOW_NOT_FOUND = 14001,
  WORKFLOW_INACTIVE = 14002,
  WORKFLOW_EXECUTION_FAILED = 14003,

  
  // 工作空间错误
  WORKSPACE_NOT_FOUND = 16001,
  WORKSPACE_ACCESS_DENIED = 16002,
}

// 工作流执行状态枚举
export enum WorkflowExecutionStatus {
  RUNNING = 'running',
  SUCCESS = 'success',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
  PENDING = 'pending',
}

// Coze插件数据类型
export interface CozePluginData {
  type: 'audio' | 'image' | 'video' | 'text';
  data: {
    url: string;
    title: string;
    description?: string;
    metadata?: Record<string, any>;
    timestamp: string;
  };
  clientId: string;
}

// 工作流监控数据
export interface WorkflowMonitorData {
  id: string;
  clientId: string;
  workflowId?: string;
  workflowName?: string;
  data: CozePluginData;
  receivedAt: string;
  isRead: boolean;
}

// 工作流监控状态
export interface WorkflowMonitorState {
  // 监控的工作流
  monitoredWorkflows: Array<{
    workflowId: string;
    workflowName: string;
    clientId: string;
    isActive: boolean;
    startedAt: string;
    lastDataReceived?: string;
    dataCount: number;
  }>;

  // 订阅的客户端ID列表
  subscribedClientIds: string[];

  // 接收到的数据
  monitorData: WorkflowMonitorData[];

  // 状态
  isMonitoring: boolean;
  selectedWorkflowId?: string;

  // 统计信息
  totalDataReceived: number;
  unreadDataCount: number;
}

// 工作流事件日志条目
export interface WorkflowEventLog {
  id: string;
  executeId?: string;
  workflowId: string;
  workflowName?: string;
  event: string;
  data: any;
  timestamp: string;
  level: 'info' | 'warning' | 'error' | 'success';
  message?: string;
  details?: any;
}

// 工作流流式响应事件类型
export interface WorkflowStreamEvent {
  event: 'workflow_started' | 'node_started' | 'node_finished' | 'workflow_finished' | 'error' | 'message' | 'data';
  data: any;
  conversation_id?: string;
  workflow_id?: string;
  node_id?: string;
  timestamp: string;
}

// 工作流流式执行状态
export interface WorkflowStreamState {
  isStreaming: boolean;
  events: WorkflowEventLog[];
  currentStep?: string;
  status: 'running' | 'completed' | 'failed' | 'cancelled';
  startTime?: string;
  endTime?: string;
  output?: any;
  error?: string;
}

// WebSocket 订阅请求
export interface CozeSubscribeRequest {
  clientId: string;
  workflowId?: string;
}

// WebSocket 订阅响应
export interface CozeSubscribeResponse {
  success: boolean;
  clientId: string;
  message: string;
  subscribedAt: string;
}

// Coze数据发送请求
export interface CozeSendDataRequest {
  api_base: string;
  clientId: string;
  data: any;
}

// Coze Zone 配置
export interface CozeZoneConfig {
  defaultBaseUrl: string;
  maxRetries: number;
  timeout: number;
  pollingInterval: number; // 执行状态轮询间隔（毫秒）
}

// 任务管理相关类型定义
export interface Task {
  id: string;                           // 后端生成的任务ID
  name: string;                         // 任务名称
  description?: string;                 // 任务描述
  workflowId: string;                   // 关联的工作流ID (Coze服务端)
  workflowName: string;                 // 工作流名称（冗余存储）
  workflowVersion?: number;             // 工作流版本

  // 输入输出数据
  inputParameters: Record<string, any>; // 输入参数
  outputData?: any;                     // 输出结果
  errorMessage?: string;                // 错误信息

  // 执行状态
  status: TaskStatus;                   // 任务状态
  executionStatus?: ExecutionStatus;    // 执行状态

  // 时间信息
  createdAt: string;                    // 创建时间
  updatedAt: string;                    // 更新时间
  executedAt?: string;                  // 执行时间
  completedAt?: string;                 // 完成时间

  // 执行关联
  cozeExecutionId?: string;             // Coze执行ID（执行后关联）
  cozeConversationId?: string;          // Coze会话ID

  // 元数据
  tags?: string[];                      // 标签
  priority?: 'low' | 'medium' | 'high'; // 优先级
  createdBy?: string;                   // 创建者
  metadata?: Record<string, any>;       // 扩展元数据
}

// 任务状态枚举
export enum TaskStatus {
  DRAFT = 'draft',                      // 草稿状态（未执行）
  EXECUTING = 'executing',              // 执行中
  COMPLETED = 'completed',              // 执行完成
  FAILED = 'failed',                    // 执行失败
  CANCELLED = 'cancelled',              // 已取消
}

// 执行状态枚举
export enum ExecutionStatus {
  PENDING = 'pending',                  // 等待执行
  RUNNING = 'running',                  // 运行中
  SUCCESS = 'success',                  // 成功
  FAILED = 'failed',                    // 失败
  TIMEOUT = 'timeout',                  // 超时
  CANCELLED = 'cancelled',              // 已取消
}

// 创建任务请求
export interface CreateTaskRequest {
  name?: string;                        // 任务名称（可选，默认使用工作流名称）
  description?: string;                 // 任务描述
  workflowId: string;                   // 工作流ID
  workflowName?: string;                // 工作流名称（可选）
  inputParameters: Record<string, any>; // 输入参数
  tags?: string[];                      // 标签
  priority?: 'low' | 'medium' | 'high'; // 优先级
  metadata?: Record<string, any>;       // 扩展元数据
}

// 更新任务请求
export interface UpdateTaskRequest {
  name?: string;                        // 任务名称
  description?: string;                 // 任务描述
  inputParameters?: Record<string, any>; // 输入参数
  outputData?: any;                     // 输出结果
  status?: TaskStatus;                  // 任务状态
  executionStatus?: ExecutionStatus;    // 执行状态
  errorMessage?: string;                // 错误信息
  executedAt?: string;                  // 执行时间
  completedAt?: string;                 // 完成时间
  cozeExecutionId?: string;             // Coze执行ID
  cozeConversationId?: string;          // Coze会话ID
  tags?: string[];                      // 标签
  priority?: 'low' | 'medium' | 'high'; // 优先级
  metadata?: Record<string, any>;       // 扩展元数据
}

// 执行任务请求
export interface ExecuteTaskRequest {
  taskId?: string;                      // 任务ID（可选，不携带则生成新任务）
  workflowId: string;                   // 工作流ID
  inputParameters?: Record<string, any>; // 输入参数（可选，使用任务存储的参数）
  saveAsTask?: boolean;                 // 是否保存为任务（默认true）
  taskName?: string;                    // 任务名称（保存为任务时使用）
  taskDescription?: string;             // 任务描述（保存为任务时使用）
  useStream?: boolean;                  // 是否使用流式执行（默认true）
  onEvent?: (event: WorkflowStreamEvent) => void; // 流式事件回调（仅流式执行时使用）
  signal?: AbortSignal;                 // 取消信号（仅流式执行时使用）
}

// 执行任务响应
export interface ExecuteTaskResponse {
  taskId?: string;                      // 任务ID（如果保存为任务）
  executionId: string;                  // 执行ID
  status: string;                       // 执行状态
  message?: string;                     // 消息
}

// 任务查询过滤器
export interface TaskFilter {
  workflowId?: string;                  // 工作流ID
  status?: TaskStatus;                  // 任务状态
  executionStatus?: ExecutionStatus;    // 执行状态
  tags?: string[];                      // 标签
  priority?: 'low' | 'medium' | 'high'; // 优先级
  createdAfter?: string;                // 创建时间筛选
  createdBefore?: string;               // 创建时间筛选
  limit?: number;                       // 限制数量
  offset?: number;                      // 偏移量
}

// 任务列表响应
export interface TaskListResponse {
  tasks: Task[];                        // 任务列表
  total: number;                        // 总数
  hasMore: boolean;                     // 是否有更多
}

// 任务统计信息
export interface TaskStatistics {
  total: number;                        // 总任务数
  byStatus: Record<TaskStatus, number>; // 按状态统计
  byExecutionStatus: Record<ExecutionStatus, number>; // 按执行状态统计
  byWorkflow: Record<string, number>;   // 按工作流统计
  recentExecutions: Task[];             // 最近执行的任务
}

// 扩展现有的CozeZoneTabData接口
export interface CozeZoneTabData {
  id: string;
  label: string;
  type: 'coze_zone';
  loading: boolean;
  error: string | null;

  // Coze Zone 特有字段
  accountId: string;
  workspaceId: string;

  // 数据
  accounts: CozeAccount[];
  workspaces: CozeWorkspace[];
  workflows: CozeWorkflow[];
  executions: WorkflowExecution[];
  selectedWorkflow?: CozeWorkflow;
  executionHistory: WorkflowExecution[];

  // 任务管理数据
  tasks: Task[];                        // 任务列表
  selectedWorkflowTasks: Task[];        // 当前选中工作流的任务
  taskStatistics?: TaskStatistics;      // 任务统计信息

  // 状态
  refreshing: boolean;
  executing: boolean;
  taskLoading: boolean;                 // 任务加载状态
  taskExecuting: string | null;         // 当前执行的任务ID
}

// 任务管理API类型
export interface TaskApi {
  // 获取任务列表
  getTasks(filters?: TaskFilter): Promise<TaskListResponse>;

  // 获取单个任务
  getTask(taskId: string): Promise<Task>;

  // 创建任务
  createTask(request: CreateTaskRequest): Promise<Task>;

  // 更新任务
  updateTask(taskId: string, request: UpdateTaskRequest): Promise<Task>;

  // 删除任务
  deleteTask(taskId: string): Promise<void>;

  // 执行任务
  executeTask(request: ExecuteTaskRequest): Promise<ExecuteTaskResponse>;

  // 获取任务统计
  getTaskStatistics(workflowId?: string): Promise<TaskStatistics>;
}

// 默认配置
export const DEFAULT_COZE_CONFIG: CozeZoneConfig = {
  defaultBaseUrl: 'https://api.coze.cn',
  maxRetries: 3,
  timeout: 30000,
  pollingInterval: 2000,
};